core/src/main/java/org/testcontainers/containers/image/pull/policy/ImagePullPolicy.java

```diff
@@ -0,0 +1,10 @@
+package org.testcontainers.containers.image.pull.policy;
+
+import com.github.dockerjava.api.model.Image;
+
+@FunctionalInterface
+public interface ImagePullPolicy {
+
+    boolean shouldPull(Image image);
```

Comment:
I feel uneasy about using docker-java's classes in our public APIs where otherwise possible.
Could you please add a simple interface for the `Image`? ☺️

Also, I'm not sure we should use `Image` here.
As a pull policy developer, I do not expect an instance of image available before I decide how to pull it. Maybe the name isn't correct and it should be something like `ImageRefreshPolicy`?

Reply:
Sure, I can avoid exposing Image. 
As for the name - I "borrowed" the Kubernetes term but I have no strong feelings either way. I think it makes sense to assume the policy dictates the behavior when an image exists locally, otherwise the library will always pull it from a remote registry, but I'm OK with changing it.

One thing I changed and would like your feedback on is the fact that the ctor of GenericContainer was triggering the docker pull flow immediately (by resolving the Future) - this seemed strange to me, because it means this flow would run even if the container is never used at all. Also this would've prevented me from adding the imagePullPolicy because by that time the image would have already been pulled (or not). 
So I moved this step to configure() instead - is that OK?

Reply:
@rnorth WDYT?

Reply:
I'm totally OK with moving the image pull behaviour out of the constructor. This is really quite a longstanding itch; the intention was to enable 'fail-fast' if a bad image was specified, but it's generally caused a fair amount of confusion.

👍 for changing this.